File: inf1.py
import sys
def solve(n,a,s,dp,i,j):
    if(n==1):
        return 0
    else:
        ans=0;
        if(dp[i][j]!=-1):
            ans = dp[i][j]
        else:
            dp[i][j]= max(solve(n-1,a[1:],s-a[0],dp,i+1,j),solve(n-1,a[:-1],s-a[-1],dp,i,j-1),s-max(a))
            ans=dp[i][j]
        return ans

def GetAnswer(n,a):
    dp=[[-1 for i in range(n+1)] for j in range(n+1)]
    return solve(n,a,sum(a),dp,0,n)

a=[]
n=int(input())
for i in range(n):
    a.append(sys.stdin.readline().strip)
    print(a)

s=sum(a)
print(GetAnswer(n,a))

File: inf1.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;

int solve(int n,vector<int> a,int s,vector<vector<int> >& dp, int i,int j){
     if(n==1){
        return 0;
    }
    else{
        int ans=0;
        if(dp[i][j]!=-1){
            ans = dp[i][j];
        }
        else{
            cout<<i<<" "<<j<<" ****"<<endl;
            int ma=*max_element(a.begin(),a.end());
            dp[i][j]= max(max(solve(n-1,a,s-a[i],dp,i+1,j),solve(n-1,a,s-a[j-1],dp,i,j-1)),s-ma);
            ans=dp[i][j];
        }
        return ans;
    }
}

int GetAnswer(int n,vector<int> a){
    vector<int> row(n+1,-1);
    vector<vector<int> > dp(n+1,row);
    int s=0;
    for(int i=0;i<a.size();i++){
        s+=a[i];
    }
    return solve(n,a,s,dp,0,n);
   
}
int main(){
    int test;
    int n;
    cin>>n;
    vector<int> a;
    int temp;
    for(int i=0;i<n;i++){
        cin>>temp;
        a.push_back(temp);
    }
    int ans=GetAnswer(n,a);
    cout<<ans<<endl;
}
